File: py/hako_binary/offset_map.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
import glob
import json
import sys
from hako_binary import offset_parser

class OffsetMap:
    def __init__(self, offset_path):
        self.off_path = offset_path
        self.map = {}

    def get(self, typename):
        if (typename in self.map) == False:
            filepath = self.find_filepath(self.off_path +"/*/", typename + ".offset")
            lines = offset_parser.parse_offset(filepath)
            self.map[typename] =  lines
        return self.map[typename]

    def find_filepath(self, path, filename):
        f_array = filename.split('/')
        if (len(f_array) > 1):
            filename = f_array[len(f_array) - 1]
        #print("path=" + path)
        #print("filename=" + filename)
        tmp = glob.glob(path + filename, recursive=True)
        if (len(tmp) == 0):
            print("ERROR: find_filepath(" + path + " ," + filename +")" )
            exit(1)
        return tmp[0]

def create_offmap(offset_path):
    return OffsetMap(offset_path)

File: bindings/elixir_project/lib/hako_binary/offset_map.ex
defmodule OffsetMap do
  defstruct off_path: nil, map: %{}

  def create_offmap(offset_path) do
    %OffsetMap{off_path: offset_path}
  end

  def get(%OffsetMap{off_path: off_path, map: map} = state, typename) do
    case Map.has_key?(map, typename) do
      true ->
        {Map.get(map, typename), state}

      false ->
        filepath = find_filepath("#{off_path}/*/", "#{typename}.offset")
        lines = OffsetParser.parse_offset(filepath)
        updated_map = Map.put(map, typename, lines)
        {%{state | map: updated_map}, updated_map[typename]}
    end
  end

  defp find_filepath(path, filename) do
    filepath_parts = String.split(filename, "/")
    filename = List.last(filepath_parts)

    tmp = Path.wildcard("#{path}#{filename}")

    if tmp == [] do
      IO.puts("ERROR: find_filepath(#{path}, #{filename})")
      System.halt(1)
    else
      List.first(tmp)
    end
  end
end
